src/test/java/com/microsoft/aad/msal4j/RequestThrottlingTest.java

```diff
@@ -0,0 +1,168 @@
+        httpResponse.addHeaders(headers);
+
+        EasyMock.expect(httpClientMock.send(anyObject())).andReturn(httpResponse).times(1);
+
+        PublicClientApplication app = getPublicClientApp(httpClientMock);
+
+        PowerMock.replayAll(httpClientMock);
+        return app;
+    }
+
+    private void throttlingTest(TokenEndpointResponseType tokenEndpointResponseType) throws Exception {
+        ThrottlingCache.clear();
+        // request #1 to token endpoint
+        // response contains Retry-After header
+        PublicClientApplication
+                app = getClientApplicationMockedWithOneTokenEndpointResponse(tokenEndpointResponseType);
+        try {
+            app.acquireToken(getAcquireTokenApiParameters("scope1")).join();
+        } catch (Exception ex) {
+            if (!(ex.getCause() instanceof MsalServiceException)) {
+                Assert.fail("Unexpected exception");
+            }
+        }
+        PowerMock.verifyAll();
+        PowerMock.resetAll();
+
+        // repeat same request #1, should be throttled
+        try {
+            app = getPublicClientApp();
+            app.acquireToken(getAcquireTokenApiParameters("scope1")).join();
+        } catch (Exception ex) {
+            if (!(ex.getCause() instanceof MsalThrottlingException)) {
+                Assert.fail("Unexpected exception");
+            }
+        }
+
+        // request #2 (different scope) should not be throttled
+        app = getClientApplicationMockedWithOneTokenEndpointResponse(tokenEndpointResponseType);
+        try {
+            app.acquireToken(getAcquireTokenApiParameters("scope2")).join();
+        } catch (Exception ex) {
+            if (!(ex.getCause() instanceof MsalServiceException)) {
+                Assert.fail("Unexpected exception");
+            }
+        }
+        PowerMock.verifyAll();
+        PowerMock.resetAll();
+
+        // repeat request #1, should not be throttled after
+        // throttling for this request got expired
+        Thread.sleep(THROTTLE_IN_SEC * 1000);
+        app = getClientApplicationMockedWithOneTokenEndpointResponse(tokenEndpointResponseType);
+        try {
+            app.acquireToken(getAcquireTokenApiParameters("scope1")).join();
+        } catch (Exception ex) {
+            if (!(ex.getCause() instanceof MsalServiceException)) {
+                Assert.fail("Unexpected exception");
+            }
+        }
+        PowerMock.verifyAll();
+        PowerMock.resetAll();
+    }
+
+    @Test
+    public void STSResponseContainsRetryAfterHeader_repeatedRequestsThrottled() throws Exception {
+        throttlingTest(TokenEndpointResponseType.RETRY_AFTER_HEADER);
+    }
+
+    @Test
+    public void STSResponseContainsStatusCode429_repeatedRequestsThrottled() throws Exception {
+        throttlingTest(TokenEndpointResponseType.STATUS_CODE_429);
+    }
+
+    @Test
+    public void STSResponseContainsStatusCode500_repeatedRequestsThrottled() throws Exception  {
```

Comment:
Sasha wrote 21 acceptance tests. I was thinking it is too much, but there are subtle differences in each one. It's also a guarantee that libraries behave the same.

Reply:
Please ensure the relevant acceptance tests are included here as that is the best way we have to ensure the code behaves as expected.

---
In reply to: [412276923](https://github.com/AzureAD/microsoft-authentication-library-for-java/pull/201#discussion_r412276923) [](ancestors = 412276923)

Reply:
Most of mentioned tests duplicated, and some not applicable for java.
Taking to account differences in api model of Msal and WAM we consider only refresh tokens grant requests as subject to UIRequired cache